### PR TITLE
Interactive EditMode for PeakEditing on the SliceViewer/PeaksViewer

### LINKED ISSUE #12710: Interactive EditMode for PeakEditing on the SliceViewer/PeaksViewer
This issue was originally [TRAC 11872](http://trac.mantidproject.org/mantid/ticket/11872)

This ticket is blocked by : 
 - [X] [TRAC11868](http://trac.mantidproject.org/mantid/ticket/11868)

Create peak centric interactive edit modes for the PeaksViewer.
* Allow peak by peak deletion
* Allow peak addition (restricted to HKL workspaces and non-integrated representations)
* Utilise and share tools with the instrument widget tools set as far as possible

 - - - - 
Keywords: vates

### PR DESCRIPTION
**Pull request by OwenArnold**
Fixes #12710. The ticket number (and hence branch numbering in trac was started prior to the migration to trac)

See issue for test instructions.

### CODE REVIEW

**Comment by OwenArnold** [comment]
http://www.mantidproject.org/ReleaseNotes_3_5_Diffraction

**Comment by martyngigg** [comment]
@OwenArnold: there are some warnings on RHEL6 to deal with

**Comment by martyngigg** [comment]
Most of the functionality looks good. I have, however, been able to make it crash.

To reproduce the crash simply erase all of the peaks on the non-integrated workspace using the erase tool and then try to add a peak by selecting the *add* tool and clicking anywhere in the view. 

Here is the stack trace that I received: 

```
Program received signal SIGSEGV, Segmentation fault.
0x00007feba168c014 in MantidQt::SliceViewer::PeakOverlayMultiCross::getOccupancyIntoView() const ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
(gdb) bt
#0  0x00007feba168c014 in MantidQt::SliceViewer::PeakOverlayMultiCross::getOccupancyIntoView() const
    () from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
#1  0x00007feba168bb96 in MantidQt::SliceViewer::PeakOverlayMultiCross::takeSettingsFrom(MantidQt::SliceViewer::PeakOverlayView const*) ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
#2  0x00007feba1666c8e in MantidQt::SliceViewer::ConcretePeaksPresenter::produceViews() ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
#3  0x00007feba166a0ea in MantidQt::SliceViewer::ConcretePeaksPresenter::addPeakAt(double, double) ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
#4  0x00007feba168b5d2 in MantidQt::SliceViewer::PeakOverlayInteractive::addPeakAt(int, int) ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidQtSliceViewer.so
#5  0x00007feb9e51e87a in QMetaObject::activate(QObject*, QMetaObject const*, int, void**) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#6  0x00007feba1a6711c in MantidQt::MantidWidgets::InputControllerPick::pickPointAt(int, int) ()
   from /home/dmn58364/Source/scm/stagingmantid/build/bin/libMantidWidgets.so
#7  0x00007feb9ecd038b in QWidget::event(QEvent*) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#8  0x00007feb9ec80e2c in QApplicationPrivate::notify_helper(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#9  0x00007feb9ec875dd in QApplication::notify(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#10 0x0000000000856119 in MantidApplication::notify(QObject*, QEvent*) ()
#11 0x00007feb9e50a4dd in QCoreApplication::notifyInternal(QObject*, QEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#12 0x00007feb9ec86d93 in QApplicationPrivate::sendMouseEvent(QWidget*, QMouseEvent*, QWidget*, QWidget*, QWidget**, QPointer<QWidget>&, bool) () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#13 0x00007feb9ecfb9eb in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#14 0x00007feb9ecfb289 in QApplication::x11ProcessEvent(_XEvent*) ()
   from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#15 0x00007feb9ed22b32 in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#16 0x00007feb94121e04 in g_main_context_dispatch () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#17 0x00007feb94122048 in ?? () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#18 0x00007feb941220ec in g_main_context_iteration () from /lib/x86_64-linux-gnu/libglib-2.0.so.0
#19 0x00007feb9e5377a1 in QEventDispatcherGlib::processEvents(QFlags<QEventLoop::ProcessEventsFlag>)
    () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#20 0x00007feb9ed22be6 in ?? () from /usr/lib/x86_64-linux-gnu/libQtGui.so.4
#21 0x00007feb9e5090af in QEventLoop::processEvents(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#22 0x00007feb9e5093a5 in QEventLoop::exec(QFlags<QEventLoop::ProcessEventsFlag>) ()
   from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#23 0x00007feb9e50eb79 in QCoreApplication::exec() () from /usr/lib/x86_64-linux-gnu/libQtCore.so.4
#24 0x00000000005323ae in main ()

```

File: Code/Mantid/MantidQt/SliceViewer/src/PeakOverlayMultiSphereFactory.cpp [hunk 1]
```diff
@@ -46,20 +50,40 @@ namespace MantidQt
...
+
+        }
+
+       }
 
       // Make the overlay widget.
-      return boost::make_shared<PeakOverlayMultiSphere>(m_plot, m_parent, physicalPeaks, this->m_peakColour, this->m_backColour);
+      return boost::make_shared<PeakOverlayMultiSphere>(presenter, m_plot, m_parent, physicalPeaks, m_plotXIndex, m_plotYIndex, this->m_peakColour, this->m_backColour);
     }
+    GCC_DIAG_ON(strict-aliasing)
```
Comment (martyngigg) [inline Code/Mantid/MantidQt/SliceViewer/src/PeakOverlayMultiSphereFactory.cpp]: gcc 4.4 still isn't happy. A local test seems to suggest that we need to remove the `GCC_DIAG_ON`  line for the suppression to work.

**Comment by martyngigg** [comment]
The crash has now been fixed so I'm happy to merge when the `RHEL6` warning is fixed.

**Comment by OwenArnold** [comment]
@martyngigg I think this is all OK now, warnings appear fixed, seems to have failed during archiving of results.

**Comment by martyngigg** [comment]
Agreed. The warnings are fixed and ci server just had an issue.

**Pull request merged**